src/lib/y2partitioner/sequences/md_controller.rb

```diff
@@ -0,0 +1,118 @@
+# encoding: utf-8
+
+# Copyright (c) [2017] SUSE LLC
+#
+# All Rights Reserved.
+#
+# This program is free software; you can redistribute it and/or modify it
+# under the terms of version 2 of the GNU General Public License as published
+# by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+# more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, contact SUSE LLC.
+#
+# To contact SUSE LLC about this file by physical or electronic mail, you may
+# find current contact information at www.suse.com.
+
+require "yast"
+require "y2storage"
+require "y2partitioner/device_graphs"
+
+module Y2Partitioner
+  module Sequences
+    # This class stores information about an MD RAID being created and takes
+    # care of updating the devicegraph when needed, so the different dialogs
+    # can always work directly on a real Md object in the devicegraph.
+    class MdController
+      include Yast::I18n
+      extend Forwardable
+
+      def_delegators :md, :md_level, :md_level=, :md_name
+
+      # @return [Y2Storage::Md] device being created
+      attr_reader :md
+
+      # Constructor
+      #
+      # @note This will create a new Md object in the devicegraph right away.
+      def initialize
+        textdomain "storage"
+        name = Y2Storage::Md.find_free_numeric_name(working_graph)
+        @md = Y2Storage::Md.create(working_graph, name)
+        @initial_name = @md.name
+      end
+
+      # Partitions that can be selected to become part of the MD array
+      #
+      # @return [Array<Y2Storage::Partition>]
+      def available_devices
+        working_graph.partitions.select { |part| available?(part) }
+      end
+
+      # Partitions that are already part of the MD array
+      #
+      # @return [Array<Y2Storage::Partition>]
+      def devices_in_md
+        md.plain_devices
+      end
+
+      # Modifies the list of devices of the Md array
+      #
+      # It removes any previous children (like filesystems) from the devices and
+      # adapts the partition id if possible.
+      #
+      # @param devices [Array<Y2Storage::BlkDevice>]
+      def devices_in_md=(devices)
+        md.devices.each { |dev| md.remove_device(dev) }
+        devices.each do |dev|
+          dev.adapted_id = Y2Storage::PartitionId::RAID if dev.is?(:partition)
+          dev = dev.encryption if dev.encrypted?
```

Comment:
@aschnell  Does it make sense to reuse the encryption when adding a partition to a MD?

Reply:
In other words, if a partition is already encrypted and the user chooses to add that partition to the RAID, should we add the encrypted device to the RAID or should we delete the encryption device and add only the plain partition?

Reply:
In theory it should work to use encrypted devices for MD RAID. But I think that was so far not supported and may not work in practice.